lib/ansible/modules/remote_management/dellemc/idrac/dellemc_idrac_firmware.py

```diff
@@ -0,0 +1,216 @@
+  returned: always
+  sample: "Successfully updated the firmware."
+update_status:
+  type: dict
+  description: Firmware Update job and progress details from the iDRAC.
+  returned: success
+  sample:
+    sample: {
+        'InstanceID': 'JID_XXXXXXXXXXXX',
+        'JobState': 'Completed',
+        'Message': 'Job completed successfully.',
+        'MessageId': 'REDXXX',
+        'Name': 'Repository Update',
+        'JobStartTime': 'NA',
+        'Status': 'Success',
+    }
+"""
+
+
+from ansible.module_utils.remote_management.dellemc.dellemc_idrac import iDRACConnection
+from ansible.module_utils.basic import AnsibleModule
+try:
+    from omsdk.sdkcreds import UserCredentials
+    from omsdk.sdkfile import FileOnShare
+    HAS_OMSDK = True
+except ImportError:
+    HAS_OMSDK = False
+
+
+def _validate_catalog_file(catalog_file_name):
+    normilized_file_name = catalog_file_name.lower()
+    if not normilized_file_name:
+        raise ValueError('catalog_file_name should be a non-empty string.')
+    elif not normilized_file_name.endswith("xml"):
+        raise ValueError('catalog_file_name should be an XML file.')
+
+
+def update_firmware(idrac, module):
+    """Update firmware from a network share and return the job details."""
+    msg = {}
+    msg['changed'] = False
+    msg['update_status'] = {}
+
+    try:
+        _validate_catalog_file(module.params['catalog_file_name'])
+    except ValueError as e:
+        module.fail_json(msg=str(e))
+
+    try:
+        upd_share = FileOnShare(remote=module.params['share_name'] + "/" + module.params['catalog_file_name'],
+                                mount_point=module.params['share_mnt'],
+                                isFolder=False,
+                                creds=UserCredentials(
+                                    module.params['share_user'],
+                                    module.params['share_pwd'])
+                                )
+
+        idrac.use_redfish = True
+        if '12' in idrac.ServerGeneration or '13' in idrac.ServerGeneration:
+            idrac.use_redfish = False
+
+        apply_update = True
+        msg['update_status'] = idrac.update_mgr.update_from_repo(upd_share,
+                                                                 apply_update,
+                                                                 module.params['reboot'],
+                                                                 module.params['job_wait'])
+    except RuntimeError as e:
+        module.fail_json(msg=str(e))
+
+    if "Status" in msg['update_status']:
+        if msg['update_status']['Status'] == "Success":
+            if module.params['job_wait']:
+                msg['changed'] = True
+        else:
+            err_msg = "Failed to update firmware."
+            if msg['update_status']['Message']:
+                err_msg = "{0} Message: {1}".format(err_msg, msg['update_status']['Message'])
+            module.fail_json(msg=err_msg)
```

Comment:
Maybe still return update_status in this case? It's probably simplify debug.
```suggestion
            module.fail_json(msg=err_msg, update_status=msg['update_status'])
```

Reply:
sure.. definitely this will provide more debug details beyond just error message

Reply:
JFYI you can accept such "change suggestions" right from GitHub UI :)